### DISCUSSION TITLE
Purpose of "defineLegacyParams"

### DISCUSSION THREAD

**Discussion by ke7kto** [idaholab/moose#17588] [Q&A General] [answered]
I am making my own simple class based on some of the classes in the framework (basically just mashing together the SplineFunction and PiecewiseTabular classes to try to get splines based on a data file), and I noticed the line 

`defineLegacyParams(<ClassName>)`

I thought I heard that the `defineLegacyParams` could be avoided in new code, so I tried removing it. This led to a linker error, with the compiler apparently not actually compiling `validParams` from the template declaration. Re-inserting the `defineLegacyParams` allowed my app to compile successfully. Eliminating both the template declaration and `defineLegacyParams` also compiles successfully.

What is the purpose of `defineLegacyParams` and the template declaration? Are they both essential, or can they both be avoided?

**Comment by loganharbour** [comment, marked as answer]
The old valid parameters used to be stored in the open namespace as a template specialization. The new valid parameters sit in a public, static member function. When you are creating new objects, there is no need to use the `defineLegacyParams()` macro. It is only used for backwards compatibility for folks that derive from the framework objects and do not use the new valid parameters syntax (public, static member function).

Let's say you have an object, `MyObject` that derives from `SomeBase`. You should have in the header:
```
#include "SomeBase.h"

class MyObject : public SomeBase
{
public:
  static InputParameters validParams();

  <rest of class>
};
```

and in the source:

```
#include "MyObject.h"

InputParameters
MyObject::validParams()
{
  auto params = SomeBase::validParams();
  <add additional params>
  return params;
}

<continuation of class>
```

and that's it for the InputParameter declaration.

Can you share the relevant portion of your object that is failing with the case where you said you "tried removing it"?

**Reply by ke7kto** [reply]
I can do that

header file 

    class SplineCSVFunction;

    //This section requires declareLegacyParams in the .C file 
    template <>
    InputParameters validParams<SplineCSVFunction>();
    //End section

    /**
     * Function that uses spline interpolation
     */
    class SplineCSVFunction : public Function
    {
    public:
      static InputParameters validParams();
    ...

Without the `declareLegacyParams` macro called in the .C file, the following output results:

    Using HIT from /home/me/source/moose/framework/contrib/hit
    Creating Unity /home/me/Documents/MooseProjects/per_capsule/build/unity_src/functions_Unity.C
    Compiling C++ (in opt mode) /home/me/Documents/MooseProjects/per_capsule/build/unity_src/functions_Unity.C...
    Linking Library /home/me/Documents/MooseProjects/per_capsule/lib/libper_capsule-opt.la...
    Linking Executable /home/me/Documents/MooseProjects/per_capsule/per_capsule-opt...
    /bin/ld: /home/me/Documents/MooseProjects/per_capsule/lib/libper_capsule-opt.so: undefined reference to `InputParameters 
    validParams<SplineCSVFunction>()'
    collect2: error: ld returned 1 exit status
    make: *** [/home/me/source/moose/framework/app.mk:407: 
    /home/me/Documents/MooseProjects/per_capsule/per_capsule-opt] Error 1

[this stackoverflow thread](https://stackoverflow.com/questions/8752837/undefined-reference-to-template-class-constructor) was helpful to me in explaining that the compiler won't complain if there is no formal declaration of a function that matches the template, which leads to the less helpful/intuitive linker error.

**Reply by loganharbour** [reply]
From what you posted, all you need at the top of the header is:

```
/**
 * Function that uses spline interpolation
 */
class SplineCSVFunction : public Function
{
public:
  static InputParameters validParams();
...

```

Remove everything above it. That is, remove both forward declaration of the class:
```
class SplineCSVFunction;
```

and forward declaration of template:
```
template <>
InputParameters validParams<SplineCSVFunction>();
```

They are for the "legacy" valid parameters syntax.

**Comment by dschwen** [comment]
I think it's about time to remove all that legacy params stuff

**Reply by loganharbour** [reply]
Is that you offering to do it? :-) Nonetheless, I agree. But, we do have _multiple_ applications we test that still use it...
